### PR TITLE
Git styles

### PR DESCRIPTION
**Pull request by mohuk**
- Branches
- Intro to workflow

### CODE REVIEW

**Comment by safdar-tp** [comment]
Please add this note:
1. Please set the develop branch as the default.
2. Restrict `Master` branch. Only Sr. Dev/TLs can merge/push.

**Comment by mohuk** [comment]
@safdar-tp will add note about default branch. 
Github or Git does not allow restrictions on branches but a general understanding within a team can be developed about the Senior folks being owners of `staging` or `master`.

**Comment by umayr** [comment]
You should also add _How to structure a commit message_ in the git style. Commit messages are really important in order to create a useful revision history, teams should first agree on a commit message convention that defines at least the following three things:
- **Style** Markup syntax, wrap margins, grammar, capitalization, punctuation.
- **Content** What kind of information should the body of the commit message (if any) contain? What should it not contain?
- **Metadata** How should issue tracking IDs, pull request numbers, etc. be referenced?

I'd suggest imperative commit messages which has been a thing since the beginning of the git. (Relevant: [Link](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html), [Link](https://www.git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#Commit-Guidelines), [Link](https://github.com/torvalds/subsurface/blob/master/README#L82-109), [Link](http://who-t.blogspot.co.at/2009/12/on-commit-messages.html), [Link](https://github.com/erlang/otp/wiki/writing-good-commit-messages), [Link](https://github.com/spring-projects/spring-framework/blob/30bce7/CONTRIBUTING.md#format-commit-messages))

To summarize, there are seven rules to draft a sensible commit message:
- [Separate subject from body with a blank line](http://chris.beams.io/posts/git-commit/#separate)
- [Limit the subject line to 50 characters](http://chris.beams.io/posts/git-commit/#limit-50)
- [Capitalize the subject line](http://chris.beams.io/posts/git-commit/#capitalize)
- [Do not end the subject line with a period](http://chris.beams.io/posts/git-commit/#end)
- [Use the imperative mood in the subject line](http://chris.beams.io/posts/git-commit/#imperative)
- [Wrap the body at 72 characters](http://chris.beams.io/posts/git-commit/#wrap-72)
- [Use the body to explain what and why vs. how](http://chris.beams.io/posts/git-commit/#why-not-how)

An example commit message taken from [here](http://chris.beams.io/posts/git-commit/):

```
Summarize changes in around 50 characters or less

More detailed explanatory text, if necessary. Wrap it to about 72
characters or so. In some contexts, the first line is treated as the
subject of the commit and the rest of the text as the body. The
blank line separating the summary from the body is critical (unless
you omit the body entirely); various tools like `log`, `shortlog`
and `rebase` can get confused if you run the two together.

Explain the problem that this commit is solving. Focus on why you
are making this change as opposed to how (the code explains that).
Are there side effects or other unintuitive consequenses of this
change? Here's the place to explain them.

Further paragraphs come after blank lines.

 - Bullet points are okay, too

 - Typically a hyphen or asterisk is used for the bullet, preceded
   by a single space, with blank lines in between, but conventions
   vary here

If you use an issue tracker, put references to them at the bottom,
like this:

Resolves: #123
See also: #456, #789
```

File: git/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+### Table of contents
+1.	[Branches](#branches)
+
+#### Branches
+A project should at least have the following branches
+- [master](#master)
+- [staging](#staging)
+- [develop](#develop)
+- [feature](#feature)
```
Comment (umayr) [inline git/README.md:11]: `feature` shouldn't be a branch itself, its kind of a type of branch you'd have if you're following topical branches model.

File: git/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+##### develop
+Develop branch is the main point of reference for developers and will host the source deployed on QA server. All feature branches must be branched out from develop. It's the responsibility of the developer to keep his feature branch up to date with develop. Developers should never commit directly on the develop branch instead they should submit a Pull Request to merge their feature branch into develop.
+
+##### feature
+Each feature branch belongs to one of the following and should contain the suggested naming convention:
+
+| Branch type  | Naming convention |
+| ------------- | ------------- |
+| user story  | feat/XXXX  |
+| issue  | iss/XXXX  |
```
Comment (umayr) [inline git/README.md]: `iss` doesn't make sense. Should be either `bug` or `issue`.

**Comment by umayr** [comment]
> Github or Git does not allow restrictions on branches but a general understanding within a team can be developed about the Senior folks being owners of staging or master.

@mohuk, you're correct but one can restrict `master` and `staging` from force pushes.

**Comment by mohuk** [comment]
@umayr @safdar-tp I think I found some help on [Github docs](https://help.github.com/articles/defining-the-mergeability-of-pull-requests/) and [Stackoverflow](http://stackoverflow.com/questions/7419244/elegant-solution-to-prevent-force-push-on-master-only).

:+1:

**Pull request merged**